File: src/lessons/recursion/StarEntity.java
package lessons.recursion;

import java.awt.Color;

public class StarEntity extends jlm.universe.turtles.Turtle {

	/* BEGIN SOLUTION */
	public void branch(int size) {
		forward(size);
		turnRight(360 / BRANCH_COUNT);
		forward(size);

		for (int i = 0; i < 2; i++)
			turnLeft(360 / BRANCH_COUNT);
	}

	public void star(int size, Color c) {
		setColor(c);
		for (int i = 0; i < BRANCH_COUNT; i++) {
			branch(size);
		}
	}

	public static final int BRANCH_COUNT = 5;

	public void run() {
		star(100, Color.black);
		turnRight(45);
		star(80, Color.blue);
		turnRight(45);
		star(60, Color.red);
	}
	/* END TEMPLATE */
}

File: src/lessons/recursion/star/StarEntity.scala
package lessons.recursion.star;

import java.awt.Color;

import jlm.universe.turtles.Turtle;

class ScalaStarEntity extends Turtle {

	/* BEGIN TEMPLATE */
	override def run() {
		/* BEGIN SOLUTION */
		star(100, Color.black);
		right(45);
		star(80, Color.blue);
		right(45);
		star(60, Color.red);
	}
	def branch(size:Int) {
		forward(size);
		right(360 / BRANCH_COUNT);
		forward(size);

		for (i <- 1 to 2)
			left(360 / BRANCH_COUNT);
	}

	val BRANCH_COUNT = 5;
	def star(size:Int, c:Color) {
		setColor(c);
		for (i <- 1 to BRANCH_COUNT) {
			branch(size);
		}
		/* END SOLUTION */
	}
	/* END TEMPLATE */
}
